provider/azure/internal/azureauth/interactive.go

```diff
@@ -0,0 +1,347 @@
+	}
+	if client.Sender != nil {
+		graphSpt.SetSender(client.Sender)
+	}
+	if err := graphSpt.Refresh(); err != nil {
+		return "", "", errors.Trace(err)
+	}
+
+	directoryURL, err := url.Parse(graphEndpoint)
+	if err != nil {
+		return "", "", errors.Annotate(err, "parsing identity endpoint")
+	}
+	directoryURL.Path = path.Join(directoryURL.Path, tenantId)
+	directoryClient := ad.NewManagementClient(directoryURL.String())
+	authorizationClient := authorization.NewWithBaseURI(resourceManagerEndpoint, subscriptionId)
+	directoryClient.Authorizer = graphSpt
+	authorizationClient.Authorizer = armSpt
+	authorizationClient.Sender = client.Sender
+	directoryClient.Sender = client.Sender
+	setClientInspectors(&directoryClient.Client, requestInspector, "azure.directory")
+	setClientInspectors(&authorizationClient.Client, requestInspector, "azure.authorization")
+
+	userObject, err := ad.UsersClient{directoryClient}.GetCurrentUser()
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	fmt.Fprintf(stderr, "Authenticated as %q.\n", userObject.DisplayName)
+
+	fmt.Fprintln(stderr, "Creating/updating service principal.")
+	servicePrincipalObjectId, password, err := createOrUpdateServicePrincipal(
+		ad.ServicePrincipalsClient{directoryClient},
+		subscriptionId,
+		clock,
+		newUUID,
+	)
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+
+	fmt.Fprintln(stderr, "Assigning Owner role to service principal.")
+	if err := createRoleAssignment(
+		authorizationClient,
+		subscriptionId,
+		servicePrincipalObjectId,
+		newUUID,
+	); err != nil {
+		return "", "", errors.Trace(err)
+	}
+	return jujuApplicationId, password, nil
+}
+
+func setClientInspectors(
+	client *autorest.Client,
+	requestInspector autorest.PrepareDecorator,
+	loggingModule string,
+) {
+	logger := loggo.GetLogger(loggingModule)
+	client.ResponseInspector = tracing.RespondDecorator(logger)
+	client.RequestInspector = tracing.PrepareDecorator(logger)
+	if requestInspector != nil {
+		tracer := client.RequestInspector
+		client.RequestInspector = func(p autorest.Preparer) autorest.Preparer {
+			p = tracer(p)
+			p = requestInspector(p)
+			return p
+		}
+	}
+}
+
+func createOrUpdateServicePrincipal(
+	client ad.ServicePrincipalsClient,
+	subscriptionId string,
+	clock clock.Clock,
+	newUUID func() (utils.UUID, error),
+) (servicePrincipalObjectId, password string, _ error) {
+	passwordCredential, err := preparePasswordCredential(clock, newUUID)
+	if err != nil {
+		return "", "", errors.Annotate(err, "preparing password credential")
+	}
+
+	servicePrincipal, err := client.Create(
+		ad.ServicePrincipalCreateParameters{
+			ApplicationID:       jujuApplicationId,
+			AccountEnabled:      true,
+			PasswordCredentials: []ad.PasswordCredential{passwordCredential},
+		},
+		nil, // abort
+	)
+	if err != nil {
+		if !isMultipleObjectsWithSameKeyValueErr(err) {
+			return "", "", errors.Trace(err)
+		}
+	} else {
+		return servicePrincipal.ObjectID, passwordCredential.Value, nil
+	}
+
```

Comment:
We will only come here if there was an error that is MultipleObjectsWithSameKeyValueErr. Based on your comment, this means that service principal already exist. Would it be possible to comment each if/else to be absolutely clear under which circumstance they'd be executed?